frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1967 @@
+				amount,
+				self.reward_account(),
+			)?,
+			// The pool should always be created in such a way its in a state to bond extra, but if
+			// the active balance is slashed below the minimum bonded or the account cannot be
+			// found, we exit early.
+			BondType::Later => T::StakingInterface::bond_extra(self.bonded_account(), amount)?,
+		}
+
+		Ok(points_issued)
+	}
+
+	/// If `n` saturates at it's upper bound, mark the pool as destroying. This is useful when a
+	/// number saturating indicates the pool can no longer correctly keep track of state.
+	fn bound_check(&mut self, n: U256) -> U256 {
+		if n == U256::max_value() {
+			self.set_state(PoolState::Destroying)
+		}
+
+		n
+	}
+
+	// Set the state of `self`, and deposit an event if the state changed. State should never be set
+	// directly in in order to ensure a state change event is always correctly deposited.
+	fn set_state(&mut self, state: PoolState) {
+		if self.state != state {
+			self.state = state;
+			Pallet::<T>::deposit_event(Event::<T>::StateChanged {
+				pool_id: self.id,
+				new_state: state,
+			});
+		};
+	}
+}
+
+/// A reward pool.
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct RewardPool<T: Config> {
+	/// The balance of this reward pool after the last claimed payout.
+	pub balance: BalanceOf<T>,
+	/// The total earnings _ever_ of this reward pool after the last claimed payout. I.E. the sum
+	/// of all incoming balance through the pools life.
+	///
+	/// NOTE: We assume this will always be less than total issuance and thus can use the runtimes
+	/// `Balance` type. However in a chain with a burn rate higher than the rate this increases,
+	/// this type should be bigger than `Balance`.
+	pub total_earnings: BalanceOf<T>,
+	/// The total points of this reward pool after the last claimed payout.
+	pub points: RewardPoints,
+}
+
+impl<T: Config> RewardPool<T> {
+	/// Mutate the reward pool by updating the total earnings and current free balance.
+	fn update_total_earnings_and_balance(&mut self, id: PoolId) {
+		let current_balance = Self::current_balance(id);
+		// The earnings since the last time it was updated
+		let new_earnings = current_balance.saturating_sub(self.balance);
+		// The lifetime earnings of the of the reward pool
+		self.total_earnings = new_earnings.saturating_add(self.total_earnings);
+		self.balance = current_balance;
+	}
+
+	/// Get a reward pool and update its total earnings and balance
+	fn get_and_update(id: PoolId) -> Option<Self> {
+		RewardPools::<T>::get(id).map(|mut r| {
+			r.update_total_earnings_and_balance(id);
+			r
+		})
+	}
+
+	/// The current balance of the reward pool. Never access the reward pools free balance directly.
+	/// The existential deposit was not received as a reward, so the reward pool can not use it.
+	fn current_balance(id: PoolId) -> BalanceOf<T> {
```

Comment:
I would have called this the `accumulated_reward` or something else to match the docs better.